### PR TITLE
Introduce more generics and more deriving

### PR DESCRIPTION
**Pull request by nox**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17209)

### CODE REVIEW

**Comment by nox** [comment]
r? @emilio 
@bors-servo try
https://treeherder.mozilla.org/#/jobs?repo=try&revision=7c246839abf029b13241e14dff2da407a0f17674

**Comment by jdm** [comment]
```
---- properties::serialization::property_declaration_block_should_serialize_correctly stdout ----
	thread 'properties::serialization::property_declaration_block_should_serialize_correctly' panicked at 'assertion failed: `(left == right)` (left: `"width: 70px; min-height: 20px; height: 20px !important; display: inline_block; overflow: auto;"`, right: `"width: 70px; min-height: 20px; height: 20px !important; display: inline-block; overflow: auto;"`)', /home/servo/buildbot/slave/linux-dev/build/tests/unit/style/properties/serialization.rs:54
```

File: components/style_derive/to_css.rs [hunk 1]
```diff
@@ -68,3 +71,40 @@ fn where_predicate(ty: syn::Ty) -> syn::WherePredicate {
...
+///
+/// If the first Camel segment is "Moz"" or "Webkit", the result string
+/// is prepended with "-".
+fn to_css_identifier(mut camel_case: &str) -> String {
+    let mut first = true;
+    let mut result = String::with_capacity(camel_case.len());
+    while let Some(segment) = split_camel_segment(&mut camel_case) {
+        if first {
+            match segment {
+                "Webkit" | "Moz" => first = false,
```
Comment (emilio) [inline components/style_derive/to_css.rs]: `!matches`?

File: components/style_derive/to_css.rs [hunk 2]
```diff
@@ -15,9 +15,12 @@ pub fn derive(input: syn::DeriveInput) -> quote::Tokens {
     }
 
     let style = synstructure::BindStyle::Ref.into();
-    let match_body = synstructure::each_variant(&input, &style, |bindings, _| {
+    let match_body = synstructure::each_variant(&input, &style, |bindings, variant| {
         if bindings.is_empty() {
-            panic!("unit variants are not yet supported");
+            let identifier = to_css_identifier(variant.ident.as_ref());
```
Comment (emilio) [inline components/style_derive/to_css.rs]: neat :)

**Review by emilio — APPROVED** [review]

**Comment by nox** [comment]
New geckotry, without the broken `display` property: https://treeherder.mozilla.org/#/jobs?repo=try&revision=0e4a90020823f5f90887e9b033a0e6558f029d2e

**Comment by nox** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=5cb715e0f2c1816b9f863b215e03e1a13235449a without `matches` which require vendoring and I don't care.

**Pull request merged**